### PR TITLE
CodeCamp #43 Support PETR in 1.1 in `projects`

### PR DESCRIPTION
**Pull request by SekiroRong**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: projects/PETR/config/petr/petr_r50dcn_gridmask_c5.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    'motorcycle', 'bicycle', 'pedestrian', 'traffic_cone'
+]
+input_modality = dict(
+    use_lidar=False,
+    use_camera=True,
+    use_radar=False,
+    use_map=False,
+    use_external=False)
+model = dict(
+    type='Petr3D',
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_r50dcn_gridmask_c5.py:25]: Petr3D -> PETR

File: projects/PETR/config/petr/petr_r50dcn_gridmask_c5.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    use_grid_mask=True,
+    img_backbone=dict(
+        type='ResNet',
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_r50dcn_gridmask_c5.py:28]: mmdet.ResNet

File: projects/PETR/config/petr/petr_r50dcn_gridmask_c5.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
+_base_ = [
+    '../../../../configs/_base_/datasets/nus-3d.py',
+    '../../../../configs/_base_/default_runtime.py'
+]
+backbone_norm_cfg = dict(type='LN', requires_grad=True)
+
+# If point cloud range is changed, the models should also change their point
+# cloud range accordingly
+point_cloud_range = [-51.2, -51.2, -5.0, 51.2, 51.2, 3.0]
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_r50dcn_gridmask_c5.py:9]: this config seems to be not complete

File: projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
+_base_ = [
+    '../../../../configs/_base_/datasets/nus-3d.py',
+    '../../../../configs/_base_/default_runtime.py',
+    '../../../../configs/_base_/schedules/cyclic-20e.py'
+]
+backbone_norm_cfg = dict(type='LN', requires_grad=True)
+custom_imports = dict(imports=['projects.PETR.mmdet3d_plugin'])
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py:7]: projects.PETR.petr

Comment (ZCMax) [reply projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py:7]: rename `mmdet3d_plugin` to `petr`

File: projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py [hunk 5]
```diff
@@ -0,0 +1,373 @@
+_base_ = [
+    '../../../../configs/_base_/datasets/nus-3d.py',
+    '../../../../configs/_base_/default_runtime.py',
+    '../../../../configs/_base_/schedules/cyclic-20e.py'
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py:4]: Can we use such format like:
```
_base_ = [
    'mmdet::_base_/datasets/coco_instance.py',
    'mmdet::_base_/schedules/schedule_1x.py',
    'mmdet::_base_/default_runtime.py'
]
```

File: projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+metainfo = dict(classes=class_names)
+
+input_modality = dict(
+    use_lidar=False,
+    use_camera=True,
+    use_radar=False,
+    use_map=False,
+    use_external=True)
+model = dict(
+    type='Petr',
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py:32]: PETR

File: projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+        type='LoadAnnotations3D',
+        with_bbox_3d=True,
+        with_label_3d=True,
+        with_attr_label=False),
+    dict(type='ObjectRangeFilter', point_cloud_range=point_cloud_range),
+    dict(type='ObjectNameFilter', classes=class_names),
+    dict(type='LidarBox3dVersionTransfrom'),
+    dict(type='AddCamInfo'),
+    dict(
+        type='ResizeCropFlipImage', data_aug_conf=ida_aug_conf, training=True),
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py:181]: Can we use the existing pipelines with `MultiviewWrapper` to achieve this pipeline?

File: projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+    dict(
+        type='GlobalRotScaleTransImage',
+        rot_range=[-0.3925, 0.3925],
+        translation_std=[0, 0, 0],
+        scale_ratio_range=[0.95, 1.05],
+        reverse_angle=False,
+        training=True),
+    dict(type='PadMultiViewImage', size_divisor=32),
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py:189]: Check wether we need this

File: projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+        pipeline=test_pipeline,
+        box_type_3d='LiDAR',
+        metainfo=metainfo,
+        test_mode=True,
+        modality=input_modality,
+        use_valid_flag=True))
+
+# lr = 0.003  # max learning rate
+
+optimizer = dict(
```
Comment (ZCMax) [inline projects/PETR/config/petr/petr_vovnet_gridmask_p4_800x320.py:264]: do we need this config?

File: projects/PETR/mmdet3d_plugin/datasets/__init__.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .transforms.transforms_3d import (AddCamInfo, GlobalRotScaleTransImage,
+                                       LidarBox3dVersionTransfrom,
+                                       NormalizeMultiviewImage,
+                                       PadMultiViewImage, ResizeCropFlipImage)
+
+__all__ = [
+    'GlobalRotScaleTransImage', 'LidarBox3dVersionTransfrom',
+    'NormalizeMultiviewImage', 'PadMultiViewImage', 'ResizeCropFlipImage',
```
Comment (ZCMax) [inline projects/PETR/mmdet3d_plugin/datasets/__init__.py:9]: Please check whether we need 'NormalizeMultiviewImage' and 'PadMultiViewImage'

File: projects/PETR/mmdet3d_plugin/datasets/builder.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (ZCMax) [inline projects/PETR/mmdet3d_plugin/datasets/builder.py:1]: Why do we need this file?

File: projects/PETR/petr/models/builder.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (ZCMax) [inline projects/PETR/petr/models/builder.py:1]: why do we need this file?

File: projects/PETR/petr/models/detectors/petr.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+        img_metas should be double nested (i.e.  list[torch.Tensor],
+        list[list[dict]]), with the outer list indicating test time
+        augmentations.
+        """
+        if mode == 'loss':
+            return self.forward_train(**kwargs)
+        elif mode == 'predict':
+            return self.forward_test(**kwargs)
+
+    def forward_train(self,
```
Comment (ZCMax) [inline projects/PETR/petr/models/detectors/petr.py:134]: why do we have `forward_train`, it should be replaced by `loss`

File: projects/PETR/petr/models/detectors/petr.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+
+        Note this setting will change the expected inputs. When
+        `return_loss=True`, img and img_metas are single-nested (i.e.
+        torch.Tensor and list[dict]), and when `resturn_loss=False`, img and
+        img_metas should be double nested (i.e.  list[torch.Tensor],
+        list[list[dict]]), with the outer list indicating test time
+        augmentations.
+        """
+        if mode == 'loss':
+            return self.forward_train(**kwargs)
```
Comment (ZCMax) [inline projects/PETR/petr/models/detectors/petr.py:130]: `forward_train` and `forward_test` have been deprecated in 1.1 branch

File: projects/PETR/petr/models/task_modules/__init__.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmdet.models.task_modules import AssignResult, BaseAssigner
```
Comment (ZCMax) [inline projects/PETR/petr/models/task_modules/__init__.py:2]: remove this import from model

File: projects/PETR/petr/models/task_modules/assigners/__init__.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from mmdet.models.task_modules import (AssignResult, BaseAssigner,
```
Comment (ZCMax) [inline projects/PETR/petr/models/task_modules/assigners/__init__.py:2]: remove the import from mmdet

File: projects/PETR/petr/models/task_modules/builder.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import warnings
```
Comment (ZCMax) [inline projects/PETR/petr/models/task_modules/builder.py:2]: remove this file

File: projects/PETR/petr/structures/utils.py [hunk 18]
```diff
@@ -0,0 +1,399 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (ZCMax) [inline projects/PETR/petr/structures/utils.py:1]: this file already existed in `mmdet3d/structures/bbox_3d/utils.py'

**Review by ZCMax — APPROVED** [review]

File: petr [hunk 19]
```diff
@@ -0,0 +1,2 @@
+[dev-1.x-project aca498fd] petr3d-
```
Comment (ZwwWayne) [inline petr:1]: this file should be deleted

**Review by ZwwWayne — COMMENTED** [review]
we do not need more folders, all files can be put in one folder, so that we can remove these unnecessary __init__.py files.
